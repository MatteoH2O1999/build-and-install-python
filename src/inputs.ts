// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import * as core from '@actions/core';
import {InputNames} from './constants';
import {getVersionInputFromFile} from 'setup-python/src/utils';
import os from 'os';
import semverValidRange from 'semver/ranges/valid';

export enum PythonType {
  CPython = 'cpython',
  PyPy = 'pypy'
}

export class PythonVersion {
  readonly type: PythonType;
  readonly version: string;

  constructor(pythonVersion: string) {
    core.debug(`Parsing version string "${pythonVersion}"...`);
    pythonVersion = pythonVersion.toLowerCase().trim();
    if (pythonVersion.length === 0) {
      pythonVersion = 'x';
    }
    while (pythonVersion.split('.').length < 3) {
      pythonVersion = pythonVersion.concat('.x');
    }
    core.debug(`Full semver range string: "${pythonVersion}".`);
    core.debug('Checking Python type (CPython or PyPy).');
    if (pythonVersion.includes('pypy')) {
      this.type = PythonType.PyPy;
      pythonVersion = pythonVersion.replace('pypy', '').replace('-', '');
    } else {
      this.type = PythonType.CPython;
    }
    core.debug(`Python type: "${this.type}".`);
    if (pythonVersion.startsWith('v')) {
      pythonVersion = pythonVersion.replace('v', '');
    }
    if (semverValidRange(pythonVersion) === null) {
      throw new Error(
        `An invalid semver string was supplied. Got "${pythonVersion}".`
      );
    }
    this.version = pythonVersion;
    core.debug(`Final resolved version range: "${this.version}".`);
  }
}

export enum BuildBehavior {
  Allow = 'allow',
  Info = 'info',
  Warn = 'warn',
  Error = 'error',
  Force = 'force'
}

export type ActionInputs = {
  version: PythonVersion;
  architecture: string;
  cache: boolean;
  buildBehavior: BuildBehavior;
  token: string;
  checkLatest: boolean;
  allowPrereleases: boolean;
};

async function getBehavior(): Promise<BuildBehavior> {
  const behaviorInput = core
    .getInput(InputNames.ALLOW_BUILD)
    .toLowerCase()
    .trim();
  core.debug(`Parsing build behavior string ${behaviorInput}...`);
  if (behaviorInput === 'error') {
    core.debug('Found behavior "error".');
    return BuildBehavior.Error;
  }
  if (behaviorInput === 'warn') {
    core.debug('Found behavior "warn".');
    return BuildBehavior.Warn;
  }
  if (behaviorInput === 'info') {
    core.debug('Found behavior "info".');
    return BuildBehavior.Info;
  }
  if (behaviorInput === 'allow') {
    core.debug('Found behavior "allow".');
    return BuildBehavior.Allow;
  }
  if (behaviorInput === 'force') {
    core.debug('Found behavior "force".');
    return BuildBehavior.Force;
  }
  throw new Error(
    `Unrecognized input value for "${
      InputNames.ALLOW_BUILD
    }". Expected one of "${BuildBehavior.Allow}", "${BuildBehavior.Info}", "${
      BuildBehavior.Warn
    }", "${BuildBehavior.Error}, ${BuildBehavior.Force}". Got "${core.getInput(
      InputNames.ALLOW_BUILD
    )}".`
  );
}

async function getCache(): Promise<boolean> {
  core.debug('Parsing cache behavior...');
  try {
    return core.getBooleanInput(InputNames.CACHE_BUILD);
  } catch (error) {
    throw new Error(
      `Expected boolean value for input "${
        InputNames.CACHE_BUILD
      }". Supported values are "true", "false", "True", "False", "TRUE", "FALSE". Got "${core.getInput(
        InputNames.CACHE_BUILD
      )}".`
    );
  }
}

async function getCheckLatest(): Promise<boolean> {
  core.debug('Parsing check latest input');
  try {
    return core.getBooleanInput(InputNames.CHECK_LATEST);
  } catch (error) {
    throw new Error(
      `Expected boolean value for input "${
        InputNames.CHECK_LATEST
      }". Supported values are "true", "false", "True", "False", "TRUE", "FALSE". Got "${core.getInput(
        InputNames.CHECK_LATEST
      )}".`
    );
  }
}

async function getAllowPrereleases(): Promise<boolean> {
  core.debug('Parsing allow prereleases input');
  try {
    return core.getBooleanInput(InputNames.PRERELEASES);
  } catch (error) {
    throw new Error(
      `Expected boolean value for input "${
        InputNames.PRERELEASES
      }". Supported values are "true", "false", "True", "False", "TRUE", "FALSE". Got "${core.getInput(
        InputNames.PRERELEASES
      )}".`
    );
  }
}

async function getArchitecture(): Promise<string> {
  const archString: string = core.getInput(InputNames.ARCHITECTURE);
  core.debug(`Parsing architecture string "${archString}"`);
  if (archString === '') {
    core.debug("architecture string is ''. Using os.arch()...");
    return os.arch();
  }
  core.debug(`Using architecture "${archString}"...`);
  return archString;
}

async function extractPythonVersion(): Promise<string> {
  let pythonVersions: string[] = core.getMultilineInput(
    InputNames.PYTHON_VERSION
  );
  let pythonVersionFile: string = core.getInput(InputNames.PYTHON_VERSION_FILE);
  core.debug(
    `Parsing python-version string "${pythonVersions.join(
      ' '
    )}" and python-version-file string "${pythonVersionFile}"`
  );

  if (pythonVersions.length > 0 && pythonVersionFile !== '') {
    core.warning(
      `Both "${InputNames.PYTHON_VERSION}" and "${InputNames.PYTHON_VERSION_FILE}" inputs were supplied. Only "${InputNames.PYTHON_VERSION}" will be used.`
    );
  }

  if (pythonVersions.length === 0 && pythonVersionFile === '') {
    core.info(
      `Neither "${InputNames.PYTHON_VERSION}" and "${InputNames.PYTHON_VERSION_FILE}" inputs were supplied. Defaulting to using ".python-version" file...`
    );
    pythonVersionFile = '.python-version';
  }

  if (pythonVersions.length === 0) {
    try {
      pythonVersions = getVersionInputFromFile(pythonVersionFile);
    } catch (error) {
      if (error instanceof Error) {
        core.warning(error.message);
      }
      pythonVersions = [];
    }
  }

  let pythonVersion = '';

  if (pythonVersions.length > 0) {
    if (pythonVersions.length > 1) {
      core.warning(
        `This action does not support installing multiple versions. Will instll only ${pythonVersions[0]}`
      );
    }
    pythonVersion = pythonVersions[0];
  }

  core.debug(`Using python version string "${pythonVersion}"...`);
  return pythonVersion;
}

export async function parseInputs(): Promise<ActionInputs> {
  return {
    allowPrereleases: await getAllowPrereleases(),
    architecture: await getArchitecture(),
    buildBehavior: await getBehavior(),
    cache: await getCache(),
    checkLatest: await getCheckLatest(),
    token: core.getInput(InputNames.TOKEN),
    version: new PythonVersion(await extractPythonVersion())
  };
}
