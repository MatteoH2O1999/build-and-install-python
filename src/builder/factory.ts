// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import * as core from '@actions/core';
import * as github from '@actions/github';
import * as semver from 'semver';
import Builder from './builder';
import {CPythonRepo} from '../constants';
import LinuxBuilder from './linux';
import MacOSBuilder from './darwin';
import {PythonVersion} from '../inputs';
import WindowsBuilder from './windows';

export default async function getBuilder(
  version: PythonVersion,
  arch: string,
  token: string
): Promise<Builder | null> {
  const specificVersion = await getSpecificVersion(version.version, token);
  if (specificVersion === null) {
    core.info(
      `Could not resolve version range ${version.version} to any available CPython tag.`
    );
    return null;
  }
  core.info(
    `Version range ${version.version} resolved to ${specificVersion.version}. Source code uri: ${specificVersion.zipBall}`
  );
  switch (process.platform) {
    case 'win32':
      return new WindowsBuilder(specificVersion, arch);
    case 'linux':
      return new LinuxBuilder(specificVersion, arch);
    case 'darwin':
      return new MacOSBuilder(specificVersion, arch);
  }
  return null;
}

async function getSpecificVersion(
  versionRange: string,
  token: string
): Promise<PythonTag | null> {
  const tags: PythonTag[] = [];

  core.debug('Creating authenticated octokit');
  const octokit = github.getOctokit(token);

  let index = 1;
  let changed = true;
  while (changed) {
    changed = false;
    let response = null;
    core.debug(`Getting tags ${(index - 1) * 100}-${index * 100}...`);
    while (response === null) {
      try {
        response = await octokit.rest.repos.listTags({
          owner: CPythonRepo.OWNER,
          page: index,
          per_page: 100,
          repo: CPythonRepo.REPO
        });
      } catch (error) {
        core.info('Rest API rate limit reached. Retrying in 60 seconds...');
        await sleep(60);
      }
    }
    if (response.status !== 200) {
      throw new Error('Error in getting tags.');
    }
    for (const tag of response.data) {
      changed = true;
      tags.push({version: tag.name, zipBall: tag.zipball_url});
    }
    index++;
  }

  let specificVersion: PythonTag | null = null;
  for (const tag of tags) {
    const semverString = semver.valid(tag.version)?.replace('v', '');
    core.debug(`Checking tag ${tag.version}`);
    if (semverString && semver.satisfies(semverString, versionRange)) {
      if (specificVersion === null) {
        specificVersion = {version: semverString, zipBall: tag.zipBall};
      } else if (semver.gte(semverString, specificVersion.version)) {
        specificVersion = {version: semverString, zipBall: tag.zipBall};
      }
    }
  }
  return specificVersion;
}

async function sleep(seconds: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}

export type PythonTag = {
  version: string;
  zipBall: string;
};
