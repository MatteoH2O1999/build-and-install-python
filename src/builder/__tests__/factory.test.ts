// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2025 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import {describe, expect, jest, test} from '@jest/globals';
import {failingTests, nullTests, tests} from './factory.fixtures';
import LinuxBuilder from '../linux';
import MacOSBuilder from '../darwin';
import {PythonType} from '../../inputs';
import WindowsBuilder from '../windows';
import getBuilder from '../factory';

jest.mock('@actions/core');
jest.mock('../tags.json', () => {
  const actual: unknown[] = jest.requireActual('../tags.json');
  actual.reverse();
  return {
    __esModule: true,
    default: actual
  };
});

function detectExpectedClass():
  | typeof LinuxBuilder
  | typeof MacOSBuilder
  | typeof WindowsBuilder
  | null {
  if (process.platform === 'linux') {
    return LinuxBuilder;
  } else if (process.platform === 'darwin') {
    return MacOSBuilder;
  } else if (process.platform === 'win32') {
    return WindowsBuilder;
  }
  return null;
}

describe('getBuilder', () => {
  const expectedBuilder = detectExpectedClass();

  test.each(tests)(
    `Requesting Python $inputVersion for architecture $inputArchitecture leads to a ${
      expectedBuilder?.name || 'null'
    } with specific version $expectedSpecificVersion for architecture $expectedArchitecture`,
    async ({
      inputVersion,
      inputArchitecture,
      expectedSpecificVersion,
      expectedArchitecture
    }) => {
      const builder = await getBuilder(inputVersion, inputArchitecture);

      expect(builder).toBeInstanceOf(expectedBuilder);
      expect(builder?.specificVersion).toEqual(expectedSpecificVersion);
      expect(builder?.arch).toEqual(expectedArchitecture);
    }
  );

  test.each(failingTests)(
    'Requesting Python %s leads to an error',
    async pythonVersion => {
      await expect(
        getBuilder(pythonVersion, process.arch)
      ).rejects.toThrowErrorMatchingSnapshot();
    }
  );

  test.each(nullTests)(
    'Requesting Python %s leads to a null builder',
    async pythonVersion => {
      const builder = await getBuilder(pythonVersion, 'x64');

      expect(builder).toBeNull();
    }
  );

  test(`Requesting any CPython version from platform different from "win32", "darwin" or "linux" leads to a null builder`, async () => {
    const realProcess = process;
    global.process = {platform: 'freebsd'} as typeof process;

    const builder = await getBuilder(
      {freethreaded: false, type: PythonType.CPython, version: '3.x.x'},
      'x64'
    );

    expect(builder).toBeNull();

    global.process = realProcess;
  });
});
