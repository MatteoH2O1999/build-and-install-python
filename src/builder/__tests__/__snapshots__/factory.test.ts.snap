// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBuilder Requesting Python { type: 'pypy', version: '2.x.x' } leads to an error 1`] = `"Only CPython can be built from source. Got pypy."`;

exports[`getBuilder Requesting Python { type: 'pypy', version: '3.x.x' } leads to an error 1`] = `"Only CPython can be built from source. Got pypy."`;
